File: C/Thread/5.7.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <unistd.h>

double u = 1;
int compteur = 0;
pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER; // 

void* thread1(void* arg) 
{
    while(compteur < 10) {
        pthread_mutex_lock(&mutex);
        compteur++;
        u = (u - 1)*(u - 1)/4;
        printf("T1 -> u = %.2f, compteur = %d\n", u, compteur);
        pthread_mutex_unlock(&mutex);
        sleep(rand()%5 + 1);
    }
    return NULL;
}

void* thread2(void* arg) 
{
    while(compteur < 10) {
        pthread_mutex_lock(&mutex);
        compteur++;
        u = (u - 2)*(u - 2)/6;
        printf("T2 -> u = %.2f, compteur = %d\n", u, compteur);
        pthread_mutex_unlock(&mutex);
        sleep(rand()%5 + 1);
    }
    return NULL;
}

int main() 
{
    srand(time(NULL));
    pthread_t t1, t2;
    pthread_create(&t1, NULL, thread1, NULL);
    pthread_create(&t2, NULL, thread2, NULL);

    pthread_join(t1, NULL);
    pthread_join(t2, NULL);
    return 0;
}

File: Python/Thread/5.7.py
import threading, time, random

u = 1
compteur = 0
mutex = threading.Lock()

def f1():
    global u, compteur
    while compteur < 10:
        with mutex:
            compteur += 1
            u = (u-1)**2/4
            print(f"T1 -> u={u:.2f}, compteur={compteur}")
        time.sleep(random.randint(1,5))

def f2():
    global u, compteur
    while compteur < 10:
        with mutex:
            compteur += 1
            u = (u-2)**2/6
            print(f"T2 -> u={u:.2f}, compteur={compteur}")
        time.sleep(random.randint(1,5))

t1 = threading.Thread(target=f1)
t2 = threading.Thread(target=f2)

t1.start()
t2.start()

t1.join()
t2.join()